### PR TITLE
planner: fix incorrect TableDual plan built from nulleq

### LINKED ISSUE #24477: The result of using the plan of TableDual with the statement of NULL-safe equal is wrong
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t2(a int,b year,key(b));
insert into t values(null,null);
select * from t where b <=> null;
```
The plan is
```
+-------------+---------+------+---------------+---------------+
| id          | estRows | task | access object | operator info |
+-------------+---------+------+---------------+---------------+
| TableDual_5 | 0.00    | root |               | rows:0        |
+-------------+---------+------+---------------+---------------+
```

### 2. What did you expect to see? (Required)
```
+------+------+
| a    | b    |
+------+------+
| NULL | NULL |
+------+------+
```
### 3. What did you see instead (Required)
Empty set
### 4. What is your TiDB version? (Required)

Release Version: v4.0.0-beta.2-2807-g289dcfefd-dirty
Edition: Community
Git Commit Hash: 289dcfefdd159345972f71cf42c9d6e594255290
Git Branch: master
UTC Build Time: 2021-05-06 15:11:16
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by sylzd**
### What problem does this PR solve?

Issue Number: close #24477  

Problem Summary:

### What is changed and how it works?

What's Changed:

util/ranger/points.go

How it Works:

If `year type col` <=> null , values.ToInt64 in `refineValueAndOp` will return err.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Fix a nulleq bug

### CODE REVIEW

**Comment by sylzd** [comment]
@eurekaka  PTAL~

File: util/ranger/points.go [hunk 1]
```diff
@@ -250,6 +250,12 @@ func (r *builder) buildFormBinOp(expr *expression.ScalarFunction) []*point {
 			// If the original value is adjusted, we need to change the condition.
 			// For example, col < 2156. Since the max year is 2155, 2156 is changed to 2155.
 			// col < 2155 is wrong. It should be col <= 2155.
+
+			// If nulleq with null value, values.ToInt64 will return err
+			if value.IsNull() {
```
Comment (eurekaka) [inline util/ranger/points.go:255]: Nitpick, I prefer to put this check into line 249, i.e, `col.GetType().Tp == mysql.TypeYear && !value.IsNull()`.

**Review by eurekaka — COMMENTED** [review]
Please add a unit test for the case mentioned in the issue.

**Review by eurekaka — APPROVED** [review]
/lgtm

**Comment by sylzd** [comment]
@qw4990  @winoros   PTAL

**Review by winoros — APPROVED** [review]

**Pull request merged**